New post talking about the latests Stork features (#1836)
quarkusio/quarkusio.github.io: _posts/2023-11-23-stork-latest-news.adoc

        // and a static service registrar
        stork.defineIfAbsent("my-second-service",
                ServiceDefinition.of(new StaticConfiguration().withAddressList(example),
                        new RandomConfiguration(), new StaticRegistrarConfiguration()));
    }
}
----

It’s important to note that the choice between programmatic and declarative configuration often depends on the specific requirements and constraints of
your application.

== Service discovery and selection strategies provided as CDI beans.

Users may prefer using a framework that leverages CDI mechanism to easily manage and inject dependencies and have a more testable and maintainable code.
Stork can now do that. Starting from the 2.0.1 release, users can use the service discovery and load balancer as beans.
For that, it looks for CDI beans during the initialization in addition to the SPI providers.
It is worth mentioning that this enhancement also contributes to improving the Quarkus experience.

== New Service Discovery approaches added.

We are happy to announce a few added service discovery strategies leveraging the power of DNS and Knative from 1.2 and 1.4 releases.

> With Knative, Smallrye Stork introduces seamless service discovery through its serverless infrastructure.

Reviewer: ```suggestion
With the Knative service discovery, Smallrye Stork introduces seamless service discovery through its serverless infrastructure, even when there are no 'pod' running.
```

The Stork Knative service discovery implementation is very similar to the Kubernetes one.
Stork will ask for https://knative.dev/docs/serving/reference/serving-api/#serving.knative.dev/v1.Service[Knative services] to the cluster instead of vanilla https://kubernetes.io/docs/concepts/services-networking/service/#service-resource[Kubernetes services] used by the Kubernetes implementation.
Again, to do so, Stork uses https://github.com/fabric8io/kubernetes-client/blob/master/extensions/knative/client/src/main/java/io/fabric8/knative/client/KnativeClient.java[Fabric 8 Knative Client] which is just an extension of Fabric8 Kubernetes Client

> The DNS-based service discovery is also here to stay. When a service has registered one or more instances in a Domain Name System (DNS) server,

Reviewer: Add an extra line.

> Stork will be able to discover them by querying the DNS. This strategy is simple and widely used, so Stork could not fail to implement it.

Reviewer: Mention the record type, and provide an example.

== New Sticky service selection strategy

> The Stork load balancer family has been improved with a new one: the sticky service selection implementation.

Reviewer: ```suggestion
The Stork load balancer family has been extended with a new one: the sticky service selection implementation.
```

The sticky service selection implemented by Stork refers to a strategy where a client "sticks" to a particular instance of a service until it fails,
then it selects another one. It is also possible to configure a backoff period for specifying how long a failing service instance should be retried.
This can be useful in scenarios where maintaining a consistent connection to the same instance is beneficial, such as when dealing with sessions or
stateful applications.

> == Enhance the service instances cache expiration policy.

Reviewer: ```suggestion
== Enhanced service instances cache expiration policy.
```

> From almost the very beginning, Stork has supported in-memory caching of discovered instances through the extension of the CachingServiceDiscovery class.

Reviewer: ```suggestion
Since almost the first release, Stork has provided in-memory caching of discovered instances by extending the `CachingServiceDiscovery` class.
```

> As of version 1.3, this capability has been expanded to allow not only the retention of the cache for a specified duration but also the implementation of custom

Reviewer: ```suggestion
As of version 1.3, this capability has been expanded to allow the retention of the cached service instances for a specified duration and the implementation of custom business logic for decision-making and data expiration.
```

business logic for decision-making and data expiration.

This enhancement was driven by the specific requirements of Kubernetes service discovery as contacting the cluster too frequently can result in performance
problems. So, out of the box, Stork Kubernetes service discovery now comes with a tailored cache expiration strategy to keep service instances until an event occurs.

> 

Reviewer: Remove one empty line.

> If you would like to do so for your custom service discovery implementations, you just need to override the `cache` method.

Reviewer: Not sure I understand? 
Can you add an example?

== Observability

Observability refers to the ability to understand and gain insights into the internal workings and behaviors of a system through the analysis
of its external outputs or observations. Stork becomes observable from the 3.6.0 release. This addition brings automated observability to the forefront
of service discovery and selection providing a window into how Stork behaves in real-time. Now you can effortlessly monitor metrics such as service discovery
and selection durations and error rates.

If you're leveraging Stork within your Quarkus application, now, you can easily check and analyze metrics such as service discovery and selection

> response times and errors directly in Prometheus. Check the https://quarkus.io/version/main/guides/stork-reference#configure-stork-observability[Stork reference guide] for having the details.

Reviewer: ```suggestion
response times and errors directly in Prometheus. Check the https://quarkus.io/version/main/guides/stork-reference#configure-stork-observability[Stork reference guide] for details.
```

== Documentation

> Additionally, our documentation has also been enhanced, offering a comprehensive guide for both seasoned users and those taking their first steps with Stork.

Reviewer: ```suggestion
Additionally, our documentation has also been enhanced, offering comprehensive guides for both seasoned users and those taking their first steps with Stork.
```

To further support your exploration, there is https://www.youtube.com/watch?v=fCNwxPDGt7Q[a video] and supplementary content that show Stork's capabilities in detail, don’t hesitate to check them out.

> Don't have much time? Don't worry, we have the https://www.youtube.com/shorts/F4Gd1I1zfjs[perfect video] to understand Stork in less than 1 minute.

Reviewer: I would move that to the introduction.

---
layout: post
title: 'SmallRye Stork Unwrapped: Exploring New Features and Enhancements'
date: 2023-11-23
tags: extension smallrye-stork
synopsis: Explore in detail the latest features added to Smallrye Stork
author: aureamunoz
---
:imagesdir: /assets/images/posts/smallrye-stork
Since its initial release in January 2022, Stork has undergone significant development, introducing new features that extended its capabilities and improved developer experience.

> This blog post takes a deep dive into the evolution of SmallRye Stork beyond its initial release, providing a detailed exploration of its fresh additions.

Reviewer 2: ```suggestion
This blog post takes a deep dive into the evolution of SmallRye Stork beyond it's initial release, providing a detailed exploration of its fresh additions.
```

Reviewer 2: Right, after thinking about it once more yes that's the correct form. Non-native speaker here.

Author: No problem. Thank you again

But first, let's describe briefly what Stork can do for you.
SmallRye Stork is a client-side service discovery and selection framework.

> It provides out-of-the-box integrations with Kubernetes, Eureka, and Hashicorp Consul, as well as a few selection strategies, including round-robin.

Reviewer 3: if its a few - why not list them all ?

    }

    @Override
    public Uni<List<ServiceInstance>> fetchNewServiceInstances(List<ServiceInstance> previousInstances) {
        // Retrieve services...
        DefaultServiceInstance instance =
                new DefaultServiceInstance(ServiceInstanceIds.next(), host, port, false);
        return Uni.createFrom().item(() -> Collections.singletonList(instance));
    }

    @Override
    public Uni<List<ServiceInstance>> cache(Uni<List<ServiceInstance>> uni) {
        return uni.memoize().until(() -> invalidated.get());
    }

    //command-based cache invalidation: user triggers the action to invalidate the cache.
    public void invalidate() {
        invalidated.set(true);
    }

}
----

You can check the https://github.com/smallrye/smallrye-stork/blob/main/service-discovery/kubernetes/src/main/java/io/smallrye/stork/servicediscovery/kubernetes/KubernetesServiceDiscovery.java[Kubernetes Service Discovery code] for further details about an event-based invalidation example.

== Observability

Observability refers to the ability to understand and gain insights into the internal workings and behaviors of a system through the analysis

> of its external outputs or observations. Stork becomes observable from the 3.6.0 release. This addition brings automated observability to the forefront

Reviewer 3: ```suggestion
of its external outputs or observations. Stork becomes observable from Quarkus 3.6.0 release. This addition brings automated observability to the forefront
```

I'm assuming this refer to Quarkus versionn rather than Stork as other version numbers mentioned was 1.3 and 1.4
